crates/volta-core/src/version/mod.rs

```diff
@@ -1,84 +1,110 @@
+            VersionSpec::Exact(version) => version.fmt(f),
+            VersionSpec::Tag(tag) => tag.fmt(f),
         }
     }
 }
 
-impl Default for VersionSpec {
-    fn default() -> Self {
-        VersionSpec::Lts
+impl fmt::Display for VersionTag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            VersionTag::Latest => write!(f, "latest"),
+            VersionTag::Lts => write!(f, "lts"),
+            VersionTag::Custom(s) => s.fmt(f),
+            VersionTag::LtsRequirement(req) => req.fmt(f),
+        }
     }
 }
 
-impl VersionSpec {
-    pub fn exact(version: &Version) -> Self {
-        VersionSpec::Exact(version.clone())
-    }
-
-    pub fn parse(s: impl AsRef<str>) -> Fallible<Self> {
-        let s = s.as_ref();
-        s.parse().with_context(version_parse_error(s))
+impl Default for VersionSpec {
+    fn default() -> Self {
+        VersionSpec::None
     }
+}
 
-    pub fn parse_requirements(s: impl AsRef<str>) -> Fallible<VersionReq> {
-        parse_requirements(s.as_ref()).with_context(version_parse_error(s))
-    }
+impl FromStr for VersionSpec {
+    type Err = VoltaError;
 
-    pub fn parse_version(s: impl AsRef<str>) -> Fallible<Version> {
-        Version::parse(s.as_ref()).with_context(version_parse_error(s))
+    fn from_str(s: &str) -> Fallible<Self> {
+        if let Ok(version) = parse_version(s) {
+            Ok(VersionSpec::Exact(version))
+        } else if let Ok(req) = parse_requirements(s) {
+            Ok(VersionSpec::Semver(req))
+        } else {
+            s.parse().map(VersionSpec::Tag)
+        }
     }
 }
 
-impl FromStr for VersionSpec {
-    type Err = ReqParseError;
+impl FromStr for VersionTag {
+    type Err = VoltaError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
+    fn from_str(s: &str) -> Fallible<Self> {
         if s == "latest" {
-            return Ok(VersionSpec::Latest);
+            Ok(VersionTag::Latest)
         } else if s == "lts" {
-            return Ok(VersionSpec::Lts);
-        }
-
-        if let Ok(ref exact) = VersionSpec::parse_version(s) {
-            Ok(VersionSpec::exact(exact))
+            Ok(VersionTag::Lts)
         } else {
-            Ok(VersionSpec::Semver(parse_requirements(s)?))
+            Ok(VersionTag::Custom(s.into()))
         }
     }
 }
 
-fn version_parse_error<E, S>(version: S) -> impl FnOnce(&E) -> ErrorDetails
-where
-    E: std::error::Error,
-    S: AsRef<str>,
-{
-    let version = version.as_ref().to_string();
-    |_error: &E| ErrorDetails::VersionParseError { version }
+pub fn parse_requirements(s: impl AsRef<str>) -> Fallible<VersionReq> {
+    let s = s.as_ref();
+    serial::parse_requirements(s)
+        .with_context(|_| ErrorDetails::VersionParseError { version: s.into() })
+}
+
+pub fn parse_version(s: impl AsRef<str>) -> Fallible<Version> {
+    let s = s.as_ref();
+    s.parse()
+        .with_context(|_| ErrorDetails::VersionParseError { version: s.into() })
```

Comment:
Is this one of those weird cases where you need a temporary for lifetime stuff, such that this won't work?

```rust
s.as_ref()
    .parse()
    .with_context(|_| ErrorDetails::VersionParseError { version: s.into() })
```

Reply:
This was actually trying to avoid the code duplication of calling `as_ref()` in both places, since we can't do `s.into()` directly and need to do `s.as_ref().into()`. I'm happy with it either way, it was just easier to work with a realized `&str` value than an `impl AsRef<str>` when it was used multiple times.

Reply:
Ahhhh, got it. I don't have a strong opinion here, was just curious!